### ISSUE TITLE
Method overload picks first available when calling forward declared function

### ISSUE THREAD

**Issue by DeviousStoat** [microsoft/pyright#3504]
```py
from typing import Any, Callable, Generic, TypeVar, overload

T = TypeVar("T")
T2 = TypeVar("T2")

class C:
    ...

class Gen(Generic[T]):
    @overload
    def method(self, v: Callable[[], int]) -> "Gen[int]":
        ...

    @overload
    def method(self, v: Callable[[], T2]) -> "Gen[T2]":
        ...

    def method(self, v: Any) -> Any:
        ...

    def method2(self: "Gen[list[C]]") -> Any:
        ...

g = (
    Gen[None]()
    .method(lambda: second_func())
    .method2()  # error here
)

def second_func() -> list[C]:
    ...
```
Here is an example, it gives
```
Could not bind method "method2" because "Gen[int]" is not assignable to parameter "self"
   TypeVar "T@Gen" is invariant
   "int" is incompatible with "list[C]"
```
The `method` call seems to be matching the first overload even though `second_func` returns a list.
But revealing the type after the first method call seem to be what's expected.
```py
g = (
    Gen[None]()
    .method(lambda: second_func())
)
reveal_type(g)  # Type of "g" is "Gen[list[C]]"
```

And if we put the `second_func` declaration before `g` declaration this works as expected.
This also works as expected if we put `from __future__ import annotations` at the top of the file.

I am sorry the example is a bit specific, I didn't manage to reduce it further unfortunately. I am available if you need more information.

**Comment by erictraut** [comment]
Which version of pyright are you using? I'm not able to repro the problem as reported. No type error occurs with the above code, so I'm guessing that you're using an older version.

**Comment by DeviousStoat** [comment]
Here is the full repro on my side with outputs:
```py
# /tmp/t.py
from typing import Any, Callable, Generic, TypeVar, overload

T = TypeVar("T")
T2 = TypeVar("T2")

class C:
    ...

class Gen(Generic[T]):
    @overload
    def method(self, v: Callable[[], int]) -> "Gen[int]":
        ...

    @overload
    def method(self, v: Callable[[], T2]) -> "Gen[T2]":
        ...

    def method(self, v: Any) -> Any:
        ...

    def method2(self: "Gen[list[C]]") -> Any:
        ...

g = (
    Gen[None]()
    .method(lambda: second_func())
    .method2()  # error here
)

def second_func() -> list[C]:
    ...
```
```sh
$ pyright --version
pyright 1.1.248

$ pyright t.py
No configuration file found.
No pyproject.toml file found.
stubPath /tmp/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/tmp/t.py
  /tmp/t.py:27:6 - error: Could not bind method "method2" because "Gen[int]" is not assignable to parameter "self"
    TypeVar "T@Gen" is invariant
      "int" is incompatible with "list[C]" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
Completed in 0.574sec
```

I am very confused what's happening. Initally I identified this error as happening only when the class has a `Final` class attribute, but I get the output above without too so I am not sure. Could be worth trying:
```py
from typing import Any, Callable, Final, Generic, TypeVar, overload

T = TypeVar("T")
T2 = TypeVar("T2")

class C:
    FINAL: Final[list[str]] = ["hello"]

class Gen(Generic[T]):
    @overload
    def method(self, v: Callable[[], int]) -> "Gen[int]":
        ...

    @overload
    def method(self, v: Callable[[], T2]) -> "Gen[T2]":
        ...

    def method(self, v: Any) -> Any:
        ...

    def method2(self: "Gen[list[C]]") -> Any:
        ...

g = (
    Gen[None]()
    .method(lambda: second_func())
    .method2()  # error here
)

def second_func() -> list[C]:
    ...
```

**Comment by DeviousStoat** [comment]
I am using the default pyright configuration without any pyproject.toml file and I can reproduce in a fresh virtual environment.
There is something interesting if I add the `reveal_type` as you suggested:
```py
from typing import Any, Callable, Generic, TypeVar, overload

T = TypeVar("T")
T2 = TypeVar("T2")

class C:
    ...

class Gen(Generic[T]):
    @overload
    def method(self, v: Callable[[], int]) -> "Gen[int]":
        ...

    @overload
    def method(self, v: Callable[[], T2]) -> "Gen[T2]":
        ...

    def method(self, v: Any) -> Any:
        ...

    def method2(self: "Gen[list[C]]") -> Any:
        ...

g = reveal_type(Gen[None]().method(lambda: second_func())).method2()

def second_func() -> list[C]:
    ...
```

```sh
No configuration file found.
No pyproject.toml file found.
stubPath /tmp/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/tmp/t.py
  /tmp/t.py:24:17 - information: Type of "Gen[None]().method(lambda : second_func())" is "Gen[int]"
  /tmp/t.py:24:60 - error: Could not bind method "method2" because "Gen[int]" is not assignable to parameter "self"
    TypeVar "T@Gen" is invariant
      "int" is incompatible with "list[C]" (reportGeneralTypeIssues)
  /tmp/t.py:24:17 - information: Type of "Gen[None]().method(lambda : second_func())" is "Gen[list[C]]"
1 error, 0 warnings, 2 informations
Completed in 0.549sec
```
Seems like it is read twice.

I realise that there must be something wrong on my side. I will investigate further

**Comment by erictraut** [comment]
I just figured out what I couldn't repro. I was using a beta of Python 3.11, and this was affecting the default behaviors for forward declarations. When I switch back to Python 3.10, I'm able to repro the problem you're seeing. I'll investigate further.

**Comment by DeviousStoat** [comment]
I was able to pinpoint this to pyright 1.1.243 if that helps

**Comment by DeviousStoat** [comment]
Maybe simpler:
```py
from typing import Any, Callable, overload

class C:
    ...

class Gen():
    def method(self) -> Any:
        ...

@overload
def func(v: Callable[[], int]) -> int:
    ...

@overload
def func(v: Callable[[], list[C]]) -> Gen:
    ...

def func(v: Any) -> Any:
    ...

func(lambda: second_func()).method()

def second_func() -> list[C]:
    ...
```

**Comment by DeviousStoat** [comment]
Even simpler
```py
from typing import Any, Callable, overload

class C:
    ...

@overload
def func(v: Callable[[], int]) -> int:
    ...

@overload
def func(v: Callable[[], list[C]]) -> list[C]:
    ...

def func(v: Any) -> Any:
    ...

def f(v: list[C]):
    ...

t = func(lambda: second_func())
reveal_type(t)  # list[C]
f(t)  # no error
t.append(C())  # calling a method on it shows the error

def second_func() -> list[C]:
    ...
```

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.251, which I just published. It will also be included in the next release of pylance.
